File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Repository/v1/Mappers/HyperlinkFieldMapper.cs
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.enums;
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Mappers.Abstract;
using Microsoft.Extensions.Logging;
using System;

namespace Mavim.Manager.Api.Topic.Repository.v1.Mappers
{
    internal class HyperlinkFieldMapper : GenericHyperlinkFieldMapper<Model.ISingleValueSimpleField, ISingleHyperlinkField>
    {
        public HyperlinkFieldMapper(ILoggerFactory loggerFactory) : base(loggerFactory) { }

        /// <summary>
        /// Gets the generic mapped text service field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <returns></returns>
        protected override ISingleHyperlinkField GetGenericMappedServiceField(Model.ISingleValueSimpleField field)
        {
            if (field == null) throw new ArgumentNullException(nameof(field));

            return new SingleHyperlinkField
            {
                FieldId = field.FieldDefinition.ID,
                FieldName = field.FieldDefinition.Name,
                SetOrder = field.FieldSetDefinition.OrderNumber,
                Order = field.OrderNumber,
                FieldSetId = field.FieldSetDefinition.ID,
                SetName = field.FieldSetDefinition?.Name ?? string.Empty,
                FieldValue = Map(field.FieldValue),
                FieldValueType = FieldType.Hyperlink,
                Required = field.FieldDefinition.Required,
                Readonly = field.ReadOnly
            };
        }

        /// <summary>
        /// Gets the generic mapped text repo field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <param name="simpleField">The simple field.</param>
        /// <returns></returns>
        /// <exception cref="ArgumentNullException">field</exception>
        protected override object[] GetGenericMappedRepoField(ISingleHyperlinkField field, Model.ISimpleField simpleField = null)
        {
            return new object[] { field?.FieldValue?.OriginalString };
        }
    }
}

File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/mappers/hyperlinkfield.mapper.ts
import { Field } from '../models/field.model';
import { mapToApiFieldType, mapToFieldType } from './fieldvaluetype';
import { GenericHyperlinkFieldMapper } from './abstract/generic-hyperlink-field.mapper';
import { SingleHyperlinkField } from '../models/api/singlehyperlinkfield.model';

export class HyperlinkFieldMapper extends GenericHyperlinkFieldMapper<SingleHyperlinkField> {
	public get getUrlTypeName(): string {
		return 'hyperlink';
	}

	public getMappedField(field: SingleHyperlinkField): Field {
		return this.mapField(field);
	}

	public getMappedApiField(field: Field): SingleHyperlinkField {
		return this.mapApiField(field);
	}

	protected mapApiField(field: Field): SingleHyperlinkField {
		return {
			...field,
			fieldValueType: mapToApiFieldType(field.fieldValueType),
			data:
				field.data && field.data.length > 0
					? this.mapToUrl(field.data[0])
					: undefined
		};
	}

	protected mapField(field: SingleHyperlinkField): Field {
		return {
			...field,
			fieldValueType: mapToFieldType(field.fieldValueType),
			options: undefined,
			data: [this.mapToString(field.data)]
		} as Field;
	}
}
